s3/pkg/service/object.go

```diff
@@ -93,165 +92,131 @@ func (dr *StreamReader) Read(p []byte) (n int, err error) {
-		limitedDataReader = io.LimitReader(data, obj.Size)
+	if req.Size > 0 { // request.ContentLength is -1 if length is unknown
+		limitedDataReader = io.LimitReader(data, req.Size)
 	} else {
 		limitedDataReader = data
 	}
 
 	backendName := bucket.DefaultLocation
-	if obj.Location != "" {
-		backendName = obj.Location
+	if req.Location != "" {
+		backendName = req.Location
 	}
 	backend, err := utils.GetBackend(ctx, s.backendClient, backendName)
 	if err != nil {
 		log.Errorln("failed to get backend client with err:", err)
 		return err
 	}
 
-	log.Infoln("bucket location:", obj.Location, " backendtype:", backend.Type,
-		" endpoint:", backend.Endpoint)
-	bodyMd5 := md["md5Sum"]
+	log.Infoln("bucket location:", req.Location, " backendtype:", backend.Type, " endpoint:", backend.Endpoint)
+	bodyMd5 := req.Attrs["md5Sum"]
 	ctx = context.Background()
-	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_SIZE, obj.Size)
+	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_SIZE, req.Size)
 	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_MD5, bodyMd5)
 	sd, err := driver.CreateStorageDriver(backend.Type, backend)
 	if err != nil {
 		log.Errorln("failed to create storage. err:", err)
-		return nil
+		return err
 	}
-	res, err := sd.Put(ctx, limitedDataReader, obj)
+	res, err := sd.Put(ctx, limitedDataReader, &pb.Object{BucketName:req.BucketName, ObjectKey:req.ObjectKey})
 	if err != nil {
 		log.Errorln("failed to put data. err:", err)
 		return err
 	}
-	oid := res.ObjectId
-	bytesWritten := res.Written
+	var delObj *pb.DeleteObjectInput
+	defer s.removeObjectFromBackend(ctx, sd, delObj)
 
-	// Should metadata update failed, add `maybeObjectToRecycle` to `RecycleQueue`,
-	// so the object in Ceph could be removed asynchronously
-	//maybeObjectToRecycle := objectToRecycle{
-	//	location: cephCluster.Name,
-	//	pool:     poolName,
-	//	objectId: oid,
-	//}
-	if bytesWritten < obj.Size {
-		//RecycleQueue <- maybeObjectToRecycle
-		log.Warnf("write objects, already written(%d), total size(%d)\n", bytesWritten, obj.Size)
-	}
-	result.Md5 = res.Etag
-
-	/*if signVerifyReader, ok := data.(*signature.SignVerifyReader); ok {
-		credential, err = signVerifyReader.Verify()
-		if err != nil {
-			RecycleQueue <- maybeObjectToRecycle
-			return
-		}
-	}*/
-	// TODO validate bucket policy and fancy ACL
-	obj.ObjectId = oid
+	obj := &pb.Object{}
+	obj.BucketName = req.BucketName
+	obj.ObjectKey = req.ObjectKey
+	obj.Acl = req.Acl
+	obj.TenantId = tenantId
+	obj.UserId = userId
+	obj.ObjectId = res.ObjectId
 	obj.LastModified = time.Now().UTC().Unix()
 	obj.Etag = res.Etag
-	obj.ContentType = md["Content-Type"]
+	obj.ContentType = req.ContentType
 	obj.DeleteMarker = false
-	obj.CustomAttributes = md /* TODO: only reserve http header attr*/
+	obj.CustomAttributes = req.Attrs
 	obj.Type = meta.ObjectTypeNormal
 	obj.Tier = utils.Tier1 // Currently only support tier1
 	obj.StorageMeta = res.Meta
+	obj.Size = req.Size
+	obj.Location = req.Location
```

Comment:
The req.Location can be null, if it is null, then so obj.Location should be the default location of the bucket.

Reply:
@sfzeng  thanks, will fix it